firmware/shared/code/I2C.c

```diff
@@ -0,0 +1,244 @@
+/*
+ * I2C.c
+ *
+ *  Created on: Oct 10, 2020
+ *      Author: Shuhang Tan
+ *      Reference: https://github.com/uvic-auvic/Firmware-Motor_Controller_2.0/blob/master/src/I2C.c
+ *                 (I2C code for motor controller 2)
+ */
+
+#include "I2C.h"
+#include <string.h>
+#include <stdlib.h>
+#include "stm32f4xx.h"
+//#include "FreeRTOS.h"
+//#include "task.h"
+#include "LED.h"
+
+// I2C SDA and SCL pins
+#define _I2C_SDA_GPIO  (GPIO_Pin_9) // PC9
+#define _I2C_SCL_GPIO  (GPIO_Pin_8) // PA8
+
+// I2C sendBuffer size
+#define _I2C_BUFFER_SIZE    (10U)
+
+typedef struct
+{
+	uint8_t Buffer[_I2C_BUFFER_SIZE];
+} I2C_data_S;
+
+// Adding an idle state because the BUSY bit in SR2 is hard to deal with
+typedef enum{
+	I2C_STATE_SEND,    // Write
+	I2C_STATE_RECEIVE, // Read
+	I2C_STATE_IDLE,    // Idle
+}I2C_state_S;
+
+// Global variables and their initialization
+static I2C_data_S I2C_send_data;
+static I2C_data_S I2C_receive_data;
+static I2C_state_S I2C_state = I2C_STATE_IDLE;
+
+uint8_t slave_address = 0x0;
+
+// Length of the data to be sent or received (in bytes)
+volatile uint8_t data_length = 0;
+
+// Data counter, 0 when a process finished or not started
+volatile uint8_t data_count = 0;
+
+// Pointer to the buffers
+uint8_t* sendBufferPtr;
+uint8_t* receiveBufferPtr;
+
+void I2C_init(void)
+{
+	// GPIO Init
+	RCC_AHB1PeriphClockCmd(RCC_AHB1Periph_GPIOF, ENABLE);
+	RCC_AHB1PeriphClockCmd(RCC_AHB1Periph_GPIOC, ENABLE);
+
+	GPIO_InitTypeDef GPIO_InitStruct;
+	GPIO_InitStruct.GPIO_Pin   = GPIO_Pin_0;
+	GPIO_InitStruct.GPIO_Mode  = GPIO_Mode_AF;
+	GPIO_InitStruct.GPIO_Speed = GPIO_Speed_100MHz;
+	GPIO_InitStruct.GPIO_OType = GPIO_OType_OD;
+	GPIO_InitStruct.GPIO_PuPd  = GPIO_PuPd_UP;
+	GPIO_Init(GPIOF, &GPIO_InitStruct);
+
+	GPIO_InitStruct.GPIO_Pin   = GPIO_Pin_1;
+	GPIO_InitStruct.GPIO_Mode  = GPIO_Mode_AF;
+	GPIO_InitStruct.GPIO_Speed = GPIO_Speed_100MHz;
+	GPIO_InitStruct.GPIO_OType = GPIO_OType_OD;
+	GPIO_InitStruct.GPIO_PuPd  = GPIO_PuPd_UP;
+	GPIO_Init(GPIOF, &GPIO_InitStruct);
+
+	// Alternative function config for SDA and SCL
+	GPIO_PinAFConfig(GPIOF, GPIO_PinSource0, GPIO_AF_I2C2);
+	GPIO_PinAFConfig(GPIOF, GPIO_PinSource1, GPIO_AF_I2C2);
+
+
+	I2C_setup();
+
+	// NVIC Init
+	NVIC_InitTypeDef NVIC_Init_Struct;
+	NVIC_Init_Struct.NVIC_IRQChannel                   = I2C2_EV_IRQn;
+	NVIC_Init_Struct.NVIC_IRQChannelCmd                = ENABLE;
+	NVIC_Init_Struct.NVIC_IRQChannelPreemptionPriority = 5;
+	NVIC_Init_Struct.NVIC_IRQChannelSubPriority        = 0;
+	NVIC_Init(&NVIC_Init_Struct);
+
+	// Set priority for ISR and enable ISR
+	//NVIC_SetPriority(I2C2_EV_IRQn, 0);
+	NVIC_EnableIRQ(I2C2_EV_IRQn);
+}
+
+void I2C_setup(void)
+{
+	// I2C2 Init
+	RCC_APB1PeriphClockCmd(RCC_APB1Periph_I2C2, ENABLE);
+
+	// Reset I2C after a delay
+	// If you want to know why, it's because I2C won't work unless doing so.

```

Comment:
Check to see if there are any changes in the status registers before and after this delay. It would be nice if we could have a condition instead of an arbitrary delay. Feel free to keep this though if observing the status registers doesn't yield anything.

Reply:
I switched back to send7bitaddress function for now. I will check to make sure that function works consistently.